### ISSUE TITLE
Uncaught exception: Error: EPERM: operation not permitted, open kubectl.exe due to `Lock file is already being held`

### ISSUE THREAD

**Issue by papanito** [lensapp/lens#736]
**Describe the bug**

Sometimes, when switching the cluster I get the following error

```
Uncaught exception: 
Error: EPERM: operation not permitted, open 
.../kubectl.exe
```

![image](https://user-images.githubusercontent.com/15034009/91011153-12446d80-e5e4-11ea-8ce0-d558164c6e7d.png)

I've installed lens successfully on my windows machine and I also could successfully add clusters. I also can see cluster details, seems all fine. However sometimes, when I switch the cluster, I get the error mentioned above. After clicking ok, I can continue working

**To Reproduce**

TBD

It usually happens when I switch to another cluster, but not always

**Expected behavior**

I do not expect any error when switching the cluster

**Screenshots**

See above

**Environment (please complete the following information):**
- Lens Version:
- OS: `Windows Server 2016`
- Installation method (e.g. snap or AppImage in Linux): download and execute binary
 
**Logs:**
When you run the application executable from command line you will see some logging output. Please paste them here:
```
←[31merror←[39m: Failed to get a lock for 1.18.0
←[31merror←[39m: Lock file is already being held {"code":"ELOCKED","file":"C:\\Users\\sc146082as\\AppData\\Roaming\\Lens\\binaries\\kubectl\\1.18.0"}
←[31merror←[39m: Failed to get a lock for 1.18.0
←[31merror←[39m: Lock file is already being held {"code":"ELOCKED","file":"C:\\Users\\sc146082as\\AppData\\Roaming\\Lens\\binaries\\kubectl\\1.18.0"}
```

**Comment by simwak** [comment]
Having the same problem. It also "deletes" (0KB) the kubectl.exe after that happens.

![image](https://user-images.githubusercontent.com/1594813/92230137-deeac400-eeaa-11ea-8ba4-3f1b44d6ac5e.png)

**Comment by nevalla** [comment]
#778 probably fixed this by releasing the lock when downloading of kubectl fails.

**Comment by rdouman** [comment]
I experienced the same issue:

Environment
Windows 7
Lens 3.5.3

I resolved it (work around) by manually downloading kubectl for the given version and placing it in the above mentioned folder before starting up Lens.

**Comment by papanito** [comment]
Thanks @nevalla I will check with `3.6.0`

**Comment by UnleashSpirit** [comment]
Hi, Lens update itself this morning to 3.6.0
Got this error now, on 1.18.8 kubectl.exe

Windows 10
Lens 3.6.0

Lens is not usable, trying to downgrade to 3.5.3

EDIT
Downgrade not working neither, the message dialog is less frequent but still broken :(
But workaround, manual download, works also in 3.6.0 with kubectl.exe 1.18.8

**Comment by simwak** [comment]
With 3.6.3 it's even worse. Getting the error every few seconds (2-10 seconds). Totally bad experience as you need to click it away ervery time.

**Comment by LiangXingzhi** [comment]
> **Describe the bug**
> 
> Sometimes, when switching the cluster I get the following error
> 
> ```
> Uncaught exception: 
> Error: EPERM: operation not permitted, open 
> .../kubectl.exe
> ```
> 
> ![image](https://user-images.githubusercontent.com/15034009/91011153-12446d80-e5e4-11ea-8ce0-d558164c6e7d.png)
> 
> I've installed lens successfully on my windows machine and I also could successfully add clusters. I also can see cluster details, seems all fine. However sometimes, when I switch the cluster, I get the error mentioned above. After clicking ok, I can continue working
> 
> **To Reproduce**
> Steps to reproduce the behavior:
> 
> TBD
> 
> It usually happens when I switch to another cluster, but not always
> 
> **Expected behavior**
> 
> I do not expect any error when switching the cluster
> 
> **Screenshots**
> 
> See above
> 
> **Environment (please complete the following information):**
> 
> * Lens Version:
> * OS: `Windows Server 2016`
> * Installation method (e.g. snap or AppImage in Linux): download and execute binary
> 
> **Logs:**
> When you run the application executable from command line you will see some logging output. Please paste them here:
> 
> ```
> ←[31merror←[39m: Failed to get a lock for 1.18.0
> ←[31merror←[39m: Lock file is already being held {"code":"ELOCKED","file":"C:\\Users\\sc146082as\\AppData\\Roaming\\Lens\\binaries\\kubectl\\1.18.0"}
> ←[31merror←[39m: Failed to get a lock for 1.18.0
> ←[31merror←[39m: Lock file is already being held {"code":"ELOCKED","file":"C:\\Users\\sc146082as\\AppData\\Roaming\\Lens\\binaries\\kubectl\\1.18.0"}
> ```

https://github.com/lensapp/lens/pull/800 fixed this issue

**Comment by nevalla** [comment]
With `3.6` you can define Download directory in preferences? Does that help? Also it's possible to disable downloading kubectl binaries and define path to your own kubectl binary.

**Comment by simwak** [comment]
Setting the path to my own kubectl binary works! Setting a different path, for example the desktop does not work.

**Comment by nevalla** [comment]
> Setting the path to my own kubectl binary works! Setting a different path, for example the desktop does not work.

when you defined desktop, did you see any binaries downloaded? Trying to figure out what is the actual problem here. I thought that app will release the lock on all situations, but apparently I am wrong.

**Comment by simwak** [comment]
> when you defined desktop, did you see any binaries downloaded? Trying to figure out what is the actual problem here. I thought that app will release the lock on all situations, but apparently I am wrong.

It created the directory `Desktop\binaries\1.18.8` and the `kubectl.exe` is just 0kb in size.

**Comment by dealboy** [comment]
Hi, still happens in 3.6.7 on win10 :( 
Are there any workarounds?

**Comment by Ghebrehiywet** [comment]
The problem is with the 0KB `kubectl.exe` file. Download kubectl manually from Kubernetes stores e.g. in my case `https://storage.googleapis.com/kubernetes-release/release/v1.16.14/bin/windows/amd64/kubectl.exe` and replace 0KB file.

![image](https://user-images.githubusercontent.com/12528528/100889656-02f37c80-347d-11eb-909d-c41290c6eaea.png)

This works for me!

**Comment by simwak** [comment]
> The problem is with the 0KB `kubectl.exe` file. Download kubectl manually from Kubernetes stores e.g. in my case `https://storage.googleapis.com/kubernetes-release/release/v1.16.14/bin/windows/amd64/kubectl.exe` and replace 0KB file.
> 
> This works for me!

That worked also for me sometime. But not always and you need to override it way too often.

**Comment by fkocik** [comment]
Thanks @Ghebrehiywet : it works for me, seems to be a good workaround.

**Comment by seilorjunior** [comment]
Do you have any news about this issue?

**Comment by Nokel81** [comment]
The problem line is https://github.com/lensapp/lens/tree/67898152171cd8a9f6f927f12c47aa306105158c/src/main/kubectl.ts#L140.

A possible solution would be to notify the user in a similar manner to #2173

**Comment by m8ram** [comment]
Still happens with Lens 4.2.3

**Comment by Nokel81** [comment]
Thanks for the information.

**Comment by petermicuch** [comment]
In my case I did similar as @simwak, disabled `Download kubectl binaries matching the Kubernetes cluster version` option in Preferences and than this does not fail (optionally you can set path to your own kubectl if is is not in your `PATH` already). Of course I might not be using cluster matching kubectl version as a consequence. 

This is reproducible always on my setups (Hyper-V VMs with WinServer 2016). User that runs Lens is an admin on the machine and has no problem to create files in the destination folder. And I installed Lens only for that one user (but does not help even when I tried with `Install for everyone`). Lens version 4.2.4.

@Nokel81 - could you please explain why https://gihub.com/lensapp/lens/issues/736#issuecomment-806040167 is the problematic part and how would notification solve this issue?

**Comment by m8ram** [comment]
Upgraded to Lens 4.2.5 today.
The first startup still threw the error but if I close Lens and restart it now I no longer get the error.
Note: I never stopped/started Lens multiple times a day before. Is it possible you get this error only once each day/each boot?

Side note which may or may not be relevant: the auto upgrade failed: Lens 4.2.4 prompted me to upgrade. After I accepted it removed Lens 4.2.4 and then nothing happened. I manually downloaded the installer for 4.2.5 and installed from scratch.
See #2617

**Comment by zatricky** [comment]
In my org we have firewalls preventing the download as it must use the provided proxy configuration within Windows. We can disable the download option and pre-provide the binary, but it seems odd that a pre-provided functional kubectl.exe is destructively overwritten with a zero-byte file. Surely the download should succeed to a temporary location *before* the file is overwritten?

**Comment by petermicuch** [comment]
This problem is gone for me ever since I setup HTTP Proxy in `File->Preferences` properly. New `kubectl.exe` is downloaded and I see no errors except of this one on the startup:
![image](https://user-images.githubusercontent.com/50990028/124550784-802b0980-de31-11eb-942f-148a47a9bd9e.png)
But that one seems to have no influence on Lens functionality.

**Comment by m8ram** [comment]
Problem still occurs in 5.0.2-latest.20210705.2

**Comment by m8ram** [comment]
Maybe I'll jinx it by posting too soon but I did not get the error on the first launch of 5.1.0.
I did still see it before when launching 5.0.2 before upgrading.

**Comment by Nokel81** [comment]
> This problem is gone for me ever since I setup HTTP Proxy in `File->Preferences` properly. New `kubectl.exe` is downloaded and I see no errors except of this one on the startup:
> ![image](https://user-images.githubusercontent.com/50990028/124550784-802b0980-de31-11eb-942f-148a47a9bd9e.png)
> But that one seems to have no influence on Lens functionality.

That issue is slightly different and was fixed by #3362

**Comment by ishanrakitha** [comment]
Hi All, I'm getting the below error on Windows 11 and I'm using AWS SSO. 

**Lens version:** 2026.3.251250-latest

<img width="695" height="636" alt="Image" src="https://github.com/user-attachments/assets/f274f6ce-cc08-447a-ba46-e9ae668fb03f" />

Please advise. 

Thank you.

**Comment by clrothmann** [comment]
Hi @ishanrakitha, thanks for the report and for sharing the screenshot -- that actually helps a lot.

Quick note: the error in your screenshot is different from what this issue was originally opened for back in 2020. The original report was about `kubectl.exe` itself being locked (typically antivirus holding the binary). Your case is about `helm.exe` failing to read a temp kubeconfig that Lens writes -- different component, different code path, different likely root cause (looks like a race condition between Lens rewriting the temp kubeconfig and a concurrent helm process reading it, possibly triggered by AWS SSO token refresh).

Could you open a new issue for this specifically? That way we can track it cleanly and write release notes against the right thread. To help us triage:
1. Confirm you're on the latest release (2026.4.151333-latest -- our auto-updater rolls out gradually, you can download directly from https://lenshq.io/download)
2. How often does it happen -- every time? Only around SSO token refresh? Only when multiple helm operations run at once (e.g. release list + release history)?
3. Lens logs from `%AppData%\Lens\logs\` around the failure
4. Do you have Windows Defender or endpoint protection actively scanning Lens paths?

If opening a new issue is friction for you, let me know and I can open one on your behalf with the info above -- just tag me. Thanks for sticking with us on this one.

**Comment by clrothmann** [comment]
Hi @ishanrakitha,

Just a friendly check-in on this -- any update on your side? If you've had a chance to hit the helm + kubeconfig race again, even partial logs or timing would help. No pressure either way.

**Comment by alexrubtsov** [comment]
Hi @ishanrakitha, following up on your April report.

We've since traced this failure on another Windows machine. Lens was rewriting its temporary kubeconfig in rapid bursts while `helm.exe` was still reading it, which Windows reports as the "cannot access the file" error you saw. The bursts came from Lens re-listing Helm releases on every Secret change in the cluster.

That behaviour was fixed, and the fix is present in the latest public release. Could you please update (https://lenshq.io/download/) and let us know whether it still happens? If it does, it would be more effective if you report it as a new separate issue to keep this thread related to the `kubectl.exe` locking issue and your case deserves its own. If it does not reproduce, we would appreciate it if you let us know about that.

**Comment by ishanrakitha** [comment]
Hi, Thank you for the update. The issue is fixed with the latest release.
